### PR TITLE
Few fixes

### PR DESCRIPTION
**Pull request by jirislaby**
I packaged dg and needed these fixes.
https://build.opensuse.org/project/monitor/home:jirislaby:statica

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -29,7 +29,7 @@ add_library(LLVMpta SHARED
 	llvm/analysis/PointsTo/Globals.cpp
 )
 
-target_link_libraries(LLVMpta PTA)
+target_link_libraries(LLVMpta PTA ${llvm_libs})
```
Comment (mchalupa) [inline src/CMakeLists.txt]: Although this change makes sense, I'm getting this error when running LLVMInstrumentation (that links with `libLLVMpta.so`:

```
CommandLine Error: Option 'debug' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options
```

I suppose that every program that wants to use `libLLVMpta.so` links with LLVM, so is this a good step?

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -65,7 +65,30 @@ add_library(LLVMdg SHARED
...
+	analysis/PointsTo/PointerSubgraph.h
+	analysis/PointsTo/PointsToFlowInsensitive.h
+	DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/llvm-dg/analysis/PointsTo/)
+install(FILES
+	llvm/llvm-utils.h
+	DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/llvm-dg/llvm/)
+install(FILES
+	llvm/analysis/PointsTo/PointerSubgraph.h
+	llvm/analysis/PointsTo/PointsTo.h
+	DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/llvm-dg/llvm/analysis/PointsTo/)
```
Comment (mchalupa) [inline src/CMakeLists.txt]: Thanks, meant to do this for ages!

File: src/llvm/analysis/old/llvm-debug.h [hunk 3]
```diff
@@ -16,7 +16,7 @@
 
 #else
 
-#define DBG(...) ;
+#define DBG(...) do { } while (0)
```
Comment (mchalupa) [inline src/llvm/analysis/old/llvm-debug.h]: I hope this macro will be removed completely in the future. Anyway, this is good for now, thanks.

File: src/CMakeLists.txt (hunk 1 above)
Comment (mchalupa) [inline src/CMakeLists.txt]: Oh, I missed the comment about packaging. I guess you need it there...

Comment (jirislaby) [inline src/CMakeLists.txt]: Could you try with `target_link_libraries(LLVMpta PUBLIC PTA ${llvm_libs})`?

Comment (mchalupa) [inline src/CMakeLists.txt]: > Could you try with target_link_libraries(LLVMpta PUBLIC PTA ${llvm_libs})?

That does not work, but this does:

```cmake
target_link_libraries(LLVMpta PUBLIC PTA
                              INTERFACE ${llvm_libs})
```


> BTW I get:
> 
> LLVMinstr 
> Usage: LLVMinstr <config.json> <llvm IR> <outputFileName>
>
> which is expected :)

when running `LLVMinstr` alone I get it too, the problem appears when you load the analysis (e.g. using memsafety.json)

Please check whether the modified `taget_link_libraries` works for you. If so, either add it to this PR or let me know and I'll merge it and follow up with this fix.

Thanks :)

**Comment by jirislaby** [comment]
Does it work for you now?

**Comment by jirislaby** [comment]
It does NOT for me though :/. And INTERFACE is in fact incorrect. PUBLIC is what should be there -- it uses the libs both in headers and in the code. Crap. I don't know how to fix this.

**Comment by mchalupa** [comment]
Took a more close look and the problem is with the `LLVMSupport` library. The `-debug` is a static variable in `Debug.cpp`, so this option is added every time the support library is linked (http://llvm.org/docs/doxygen/html/Debug_8cpp_source.html#l00042)

A workaround would be to build with `NDEBUG`, but we sometimes need the debugging information.

Other solution (but I'm not sure it would be working, it would depend on the order of initialization) would be to get the registered options and delete it in the initialization of the pass (http://llvm.org/docs/doxygen/html/namespacellvm_1_1cl.html#a1838f41bf78d36caeab3a553295dbab2)

Nothing else have came to my mind...

**Comment by jirislaby** [comment]
I moved the shared libs fix to a separate branch and will create a new PR where we can track it further, apart from  the other changes.

**Comment by mchalupa** [comment]
Thank you.

**Pull request merged**